In [5]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [2]:
!python --version

Python 3.8.5


In [6]:
import sys
sys.version_info
sys.version

sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)

'3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]'

PEP 8 style guide
https://www.python.org/dev/peps/pep-0008/

### The difference between *bytes* and *str*

- Instances of *bytes* contain raw, unsigned 8-bit values (often displayed in the ASCII encoding):

- Instances of *str* contain Unicode *code points* that represent textual characters from human languages:

In [7]:
a = b'h\x65llo'
list(a)
a

[104, 101, 108, 108, 111]

b'hello'

In [8]:
a = 'a\u0300 propos'
list(a)
a 

['a', '̀', ' ', 'p', 'r', 'o', 'p', 'o', 's']

'à propos'

In [10]:
def to_str(bytes_or_str):
    if isinstance(bytes_or_str,bytes):
        value = bytes_or_str.decode('utf-8')
    else:
        value = bytes_or_str
    return value    # Instance of str

def to_bytes(bytes_or_str):
    if isinstance(bytearray,str):
        value = bytes_or_str.encode('utf-8')
    else:
        value = bytes_or_str
    return value    # Instance of bytes

In [16]:
to_str(b'h\x65llo')
to_str('hello')

to_bytes(b'h\x65llo')
to_bytes('bar')

'hello'

'hello'

b'hello'

'bar'

In [17]:
b'one' + b'two'
'one' + 'two'

b'onetwo'

'onetwo'

### Format string

#### C-style *%* operator

In [42]:
# escape character
print('%.2f%%' % 12.578)

12.58%


In [18]:
a = 0b10100101
b = 0x0c5f

print('Binary is %d, hex is %d' % (a,b))

Binary is 165, hex is 3167


In [20]:
key = 'my_var'
value = 1.234
formatted = '%-10s = %.2f' % (key, value)
print(formatted)

my_var     = 1.23


In [22]:
pantry = [
    ('avocados', 1.25),
    ('bananas', 2.5),
    ('cherries',15),
]
for i, (item,count) in enumerate(pantry):
    print('#%-3d: %-10s = %.2f' % (i, item,count))


#0  : avocados   = 1.25
#1  : bananas    = 2.50
#2  : cherries   = 15.00


In [25]:
for i, (item,count) in enumerate(pantry):
    print('#%-3d: %-10s = %.2f' % (
        i+1, 
        item.title(),
        int(round(count))))

#1  : Avocados   = 1.00
#2  : Bananas    = 2.00
#3  : Cherries   = 15.00


#### Formatting with a *dictionary* instead of a *tuple*

In [27]:
key = 'my_var'
value = 1.234

old_way = '%-10s = %.2f' % (key, value)
new_way = '%(key)-10s = %(value).2f' % {
    'key': key, 'value': value}    # Original 
reordered = '%(key)-10s = %(value).2f' % {
    'value': value, 'key': key}    # Swapped

assert old_way == new_way == reordered

In [29]:
name = 'Max'

template = '%s loves food. See %s cooks.'
before = template % (name, name)    # Tuple

template = '%(name)s loves food. See %(name)s cooks.'
after = template % {'name': name}    # Dictionary

assert before == after

In [31]:
for i, (item,count) in enumerate(pantry):
    before = '#%-3d: %-10s = %.2f' % (
        i+1, 
        item.title(),
        int(round(count)))
    
    after = '#%(loop)-3d: %(item)-10s = %(count)d' % {
        'loop': i + 1,
        'item': item.title(),
        'count': round(count),
    }
    
    assert before == after


AssertionError: 

#### The *format* Built-in and *str.format*


In [32]:
a = 1234.5678
formatted = format(a,',.2f')
formatted 

'1,234.57'

In [33]:
b = 'my string'
formatted = format(b,'^20s')    # ^ for centering
formatted 

'     my string      '

In [36]:
key = 'my_var'
value = 1.234

formatted = '{} = {}'.format(key, value)
formatted 

'my_var = 1.234'

In [40]:
key = 'my_var'
value = 1.234

formatted = '{:<20} = {:.2f}'.format(key, value)
formatted

'my_var               = 1.23'

In [43]:
# escape character
print('{} replace {{}}'.format(1.234))

1.234 replace {}


In [45]:
formatted = '{1} = {0}'.format(key, value)
formatted

'1.234 = my_var'

In [46]:
formatted = '{0} loves food. See {0} cooks.'.format(name)
formatted

'Max loves food. See Max cooks.'

#### Interpolated Format String - *f-strings* for short


In [50]:
key = 'my_var'
value = 1.234

formatted = f'{key} = {value}'
print(formatted)

my_var = 1.234


In [49]:
formatted = f'{key!r:<10} = {value:.2f}'
formatted

"'my_var'   = 1.23"

In [52]:
formatted = f'{key:<10} = {value:.2f}'
formatted
print(formatted)

'my_var     = 1.23'

my_var     = 1.23


In [54]:
f_string = f'{key:<10} = {value:.2f}'

c_tuple = '%-10s = %.2f' % (key, value)

str_args = '{:<10} = {:.2f}'.format(key, value)

str_kw = '{key:<10} = {value:.2f}'.format(key=key, value=value)

c_dict = '%(key)-10s = %(value).2f' % {'key': key, 'value': value}

assert c_tuple == c_dict == f_string
assert str_args == str_kw == f_string

In [56]:
for i, (item,count) in enumerate(pantry):
    old_style = '#%-3d: %-10s = %.2f' % (
        i+1, 
        item.title(),
        int(round(count)))
    
    new_style = '#{}: %{:<10s} = {}'.format(
        i + 1,
        item.title(),
        round(count),
    )
    
    f_string = f'#{i+1}: {item.title():<10s} = {round(count)}'
    
    assert old_style == new_style == f_string

AssertionError: 

### Write Helper Functions Instead of Complex Expressions

In [62]:
from urllib.parse import parse_qs

my_values = parse_qs('red=5&blue=0&green=',keep_blank_values=True)
print(my_values)
print(repr(my_values))

print('Red:        ', my_values.get('red'))
print('Green:      ', my_values.get('green'))
print('Opacity：    ', my_values.get('opacity'))

{'red': ['5'], 'blue': ['0'], 'green': ['']}
{'red': ['5'], 'blue': ['0'], 'green': ['']}
Red:         ['5']
Green:       ['']
Opacity：     None


### Prefer Multiple Assignment Unpacking Over Indexing

In [63]:
snake_calories = {
    'chips': 140,
    'popcorn': 80,
    'nuts': 190,
}
items = tuple(snake_calories.items())
items

(('chips', 140), ('popcorn', 80), ('nuts', 190))

In [64]:
favorite_snacks = {
    'salty': ('pretzels', 100),
    'sweet': ('cookies', 180),
    'veggie': ('carrots', 20),
}

((type1, (name1, cals1)),
 (type2, (name2, cals2)),
 (type3, (name3, cals3))) = favorite_snacks.items()

print(f'Favorite {type1} is {name1} with {cals1} calories')
print(f'Favorite {type2} is {name2} with {cals2} calories')
print(f'Favorite {type3} is {name3} with {cals3} calories')

Favorite salty is pretzels with 100 calories
Favorite sweet is cookies with 180 calories
Favorite veggie is carrots with 20 calories


In [66]:
def bubble_sort(a):
    for _ in range(len(a)):
        for i in range(1,len(a)):
            if a[i] < a[i-1]:
                a[i-1], a[i] = a[i], a[i-1]  # swap

names = ['pretzels', 'carrots', 'arugula', 'bacon']
bubble_sort(names)
names

['arugula', 'bacon', 'carrots', 'pretzels']

In [71]:
snakes = [('bacon', 350), ('donut', 240), ('muffin', 190)]

for rank, (name, calories) in enumerate(snakes,1):
    print(f'#{rank}: {name} has {calories} calories.')

#1: bacon has 350 calories.
#2: donut has 240 calories.
#3: muffin has 190 calories.


### Item 7: Prefer *enumerate* over *range*

In [80]:
from random import randint

random_bits = 0 
for i in range(32):
    if randint(0,1):
        random_bits |= 1 << i
print(bin(random_bits))

0b10100111110011001000101110001110


In [81]:
flavor_list = ['vanilla', 'chocolate', 'pecan', 'strawberry']

for i, flavor in enumerate(flavor_list):
    print(f'{i+1}: {flavor}')

1: vanilla
2: chocolate
3: pecan
4: strawberry


In [83]:
for i, flavor in enumerate(flavor_list,1):
    print(f'{i}: {flavor}')

1: vanilla
2: chocolate
3: pecan
4: strawberry


### Item 8: Use *zip* to process Iterators in parallel

In [92]:
names = ['Cecilia', 'Lise', 'Marie']
counts = [len(n) for n in names]
counts


[7, 4, 5]

('Cecilia', 7)

In [93]:
longest_name = names[counts.index(max(counts))]
(longest_name, max(counts))

('Cecilia', 7)

In [95]:
longest_name = None
max_count = 0

for name,count in zip(names, counts):
    if count > max_count:
        longest_name = name
        max_count = count 
print((longest_name, max_count))

('Cecilia', 7)


In [97]:
names.append('Roaslind')

In [98]:
import itertools

for name, count in itertools.zip_longest(names, counts):
    print(f'{name}: {count}')

Cecilia: 7
Lise: 4
Marie: 5
Roaslind: None


### Avoid *else* blocks after *for* and *while* loops

- Python has special syntax that allows *else* blocks to immediately follow *for* and *while* loop interior blocks.

- The *else* block after a loop runs only if the loop body did not encounter a break statement.

In [100]:
a = 14
b = 19

for i in range(2, min(a,b) + 1):
    print('Testing ',i)
    if (a % i == 0) and (b % i == 0):
        print('No coprime')
        break
else:
    print('Coprime')

Testing  2
Testing  3
Testing  4
Testing  5
Testing  6
Testing  7
Testing  8
Testing  9
Testing  10
Testing  11
Testing  12
Testing  13
Testing  14
Coprime


In [102]:
def coprime(a,b):
    for i in range(2, min(a,b) + 1):
        if (a % i == 0) and (b % i == 0):
            return False
    return True

assert coprime(14,49)

AssertionError: 

### Item 10 Prevent Repetition with Assignment Expressions

An assignment expression - also known as ***walrus operator (a := b)*** - is a new syntax introduced in Python 3.8 to solve a long-standing problem with the language that can cause code duplication.

This pattern of **fetching** a value, **checking** to see if it's non-zero, and then **using** it is extremely common in Python. 

**Example 1**
if count := fresh_fruit.get('lemon',0):
    make_lemonade(count)
else:
    out_of_stock()
    
**Example 2**
if (count := fresh_fruit.get('apple', 0)) >= 4:
    make_cider(count)
else:
    out_of_stock()

**Example 3**
if (count := fresh_fruit.get('banana', 0)) >= 2:
    pieces = slice_bananas(count)
    to_enjoy = make_smoothies(pieces)
elif (count := fresh_fruit.get('apple', 0)) >= 4:
    to_enjoy = make_cider(count)
elif count := fresh_fruit.get('lemon', 0):
    to_enjoy = make_lemonade(count)
else:
    to_enjoy = 'Nothing'

A strategy for improving the code reuse in this situation is to use *loop-and-a-half* idiom.

**loop-and-a-half idiom**

ottles = []
while True: # Loop
    fresh_fruit = pick_fruit()
    if not fresh_fruit: # And a half
        break
    for fruit, count in fresh_fruit.items():
        batch = make_juice(fruit, count)
        bottles.extend(batch)
        
**Walrus Assignment **

bottles = []
while fresh_fruit := pick_fruit():
    for fruit, count in fresh_fruit.items():
        batch = make_juice(fruit, count)
        bottles.extend(batch)